include/verilated_types.h

```diff
@@ -1003,12 +1004,157 @@ std::string VL_TO_STRING(const VlUnpacked<T_Value, T_Depth>& obj) {
     return obj.to_string();
 }
 
+class VlClass;  // See below
+
+//===================================================================
+// Class providing delayed deletion of garbage objects. Objects get deleted only when 'deleteAll()'
+// is called, or the deleter itself is destroyed.
+
+class VlDeleter final {
+    // MEMBERS
+    std::vector<VlClass*> m_garbage;  // Queue of objects to delete
+    VerilatedMutex m_mutex;  // Mutex protecting garbage queue
+
+public:
+    // CONSTRUCTOR
+    VlDeleter() = default;
+    ~VlDeleter() { deleteAll(); }
+
+private:
+    VL_UNCOPYABLE(VlDeleter);
+
+public:
+    // METHODS
+    // Adds a new object to the delete queue.
+    void put(VlClass* const objp) {
+        const VerilatedLockGuard lock{m_mutex};
+        m_garbage.push_back(objp);
+    }
+
+    // Deletes all queued garbage objects.
+    void deleteAll();
+};
+
+//===================================================================
+// Base class for all verilated classes. Includes a reference counter, and a pointer to the deleter
+// object that should destroy it after the counter reaches 0. This allows for easy construction of
+// VlClassRefs from 'this'. Also declares a virtual constructor, so that the object can be deleted
+// using a base pointer.
+
+class VlClass VL_NOT_FINAL {
+    // TYPES
+    template <typename T_Class>
+    friend class VlClassRef;  // Needed for access to the ref counter and deleter
+
+    // MEMBERS
+    std::atomic<size_t> m_counter{0};  // Reference count for this object
+    VlDeleter* m_deleter = nullptr;  // The deleter that will delete this object
+
+    // METHODS
+    // Atomically increments the reference counter
+    void incRefCount() { m_counter++; }
+    // Atomically decrements the reference counter. Assuming VlClassRef semantics are sound, it
+    // should never get called at m_counter == 0.
+    void decRefCount() {
+        if (!--m_counter) m_deleter->put(this);
+    }
+
+public:
+    // CONSTRUCTORS
+    VlClass() = default;
+    VlClass(const VlClass& copied) {}
+    virtual ~VlClass() {}
```

Comment:
This adds the virtual type (8 bytes) to every class, even non-virtual classes, which I don't think are there now.  This can be expensive.

Reply:
Yes, on the other hand, `std::shared_ptr` is 16 bytes (the new `VlClassRef` pointer is only 8 bytes), and incurs a 16 byte overhead per allocated object. I also tried a custom deleter, and that's even worse as it requires two allocations. See https://godbolt.org/z/oYce5WP5v.

As a side note, I think having a base virtual class could be useful for implementing a proper garbage collector (I was thinking of generating virtual methods that mark all referenced objects as reachable).

Reply:
Still wondering can we only have this virtual where needed?  Not sure what strictly needs it....

Reply:
We need it to call all destructors at deletion time. It is necessary if we have a `VlClassRef` to a base class, and in `VlDeleter` which needs to use a generic pointer type. So either a virtual destructor or a deleter function.

Reply:
I guess we'll live with this cost.